gh-120389: Add PyLong_FromInt64() and PyLong_AsInt64()

vstinner opened the pull request:
Add new functions to convert C <stdint.h> numbers from/to Python int:

* PyLong_FromInt32()
* PyLong_FromUInt32()
* PyLong_FromInt64()
* PyLong_FromUInt64()
* PyLong_AsInt32()
* PyLong_AsUInt32()
* PyLong_AsInt64()
* PyLong_AsUInt64()

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-120389 -->
* Issue: gh-120389
<!-- /gh-issue-number -->

vstinner committed:
gh-120389: Add PyLong_FromInt64() and PyLong_ToInt64()

Add new functions to convert C <stdint.h> numbers from/to Python int:

* PyLong_FromInt32()
* PyLong_FromUInt32()
* PyLong_FromInt64()
* PyLong_FromUInt64()
* PyLong_ToInt32()
* PyLong_ToUInt32()
* PyLong_ToInt64()
* PyLong_ToUInt64()

vstinner committed:
Add uint32_t to Doc/conf.py

zooba commented on Objects/longobject.c:
```diff
@@ -6681,3 +6681,84 @@ Py_ssize_t
+int PyLong_ToInt32(PyObject *obj, int32_t *value)
+{
+#if SIZEOF_INT == 4
+    int res = PyLong_AsInt(obj);
```
Why not just use `PyLong_AsNativeBytes` for each of these?

vstinner replied:
It's just old habits :-) I updated the implementation of "To" functions to use `PyLong_AsNativeBytes()`. It has two side effects:

* Use `__index__()` method if available.
* Always raise `ValueError` if the value is negative.

@serhiy-storchaka: Using `__index__()` is a good thing, no? :-)

zooba replied:
> Using `__index__()` is a good thing, no? :-)

If it's good enough for `PyLong_AsLong` then I don't see why it isn't good enough for these functions.

vstinner replied:
`PyLong_AsLong(obj)` calls `obj.__index__()`, whereas `PyLong_AsUnsignedLong(obj)` doesn't call `obj.__index__()`. I don't know the history of the `PyLong_AsUnsignedLong()` design, if there was a "design" for this function :-)

For new functions, I propose to call `obj.__index__()`.

vstinner replied:
The code now uses PyLong_AsNativeBytes().

vstinner committed:
Use PyLong_AsNativeBytes()

PyLong_ToUInt32() and PyLong_ToUInt64() can now use the __index__()
method if the object has the method.

zooba commented on Doc/c-api/long.rst:
```diff
@@ -337,6 +369,52 @@ distinguished from a number.  Use :c:func:`PyErr_Occurred` to disambiguate.
+.. c:function:: int PyLong_ToInt64(PyObject *obj, int64_t *value)
+
+   Similar to :c:func:`PyLong_ToInt32`, but return a signed C
+   :c:expr:`int64_t` representation.
```
Can we instead group the definitions together? I'm sure I saw someone say this syntax should work:

```
.. c:function::
   int PyLong_ToInt32(PyObject *obj, int32_t *value)
   int PyLong_ToInt64(PyObject *obj, int64_t *value)

   Docs for both
```

zooba reviewed (approved):
LGTM. Would like it if we don't have to have docs like "like this other function", but if that's the best we can do then I'll survive.

vstinner committed:
Group documentations

serhiy-storchaka commented on Modules/_testlimitedcapi/long.c:
```diff
@@ -1,7 +1,7 @@
 #ifndef Py_GIL_DISABLED
-   // Need limited C API 3.13 to test PyLong_AsInt()
-#  define Py_LIMITED_API 0x030d0000
+   // Need limited C API 3.14 to test PyLong_AsInt64()
```
```suggestion
   // Need limited C API 3.14 to test PyLong_ToInt64()
```

serhiy-storchaka commented on Modules/_testcapimodule.c:
```diff
@@ -4047,6 +4047,12 @@ PyInit__testcapi(void)
 
     PyModule_AddIntConstant(m, "the_number_three", 3);
     PyModule_AddIntMacro(m, Py_C_RECURSION_LIMIT);
+    PyModule_AddObject(m, "INT32_MIN", PyLong_FromInt32(INT32_MIN));
+    PyModule_AddObject(m, "INT32_MAX", PyLong_FromInt32(INT32_MAX));
+    PyModule_AddObject(m, "UINT32_MAX", PyLong_FromUInt32(UINT32_MAX));
+    PyModule_AddObject(m, "INT64_MIN", PyLong_FromInt64(INT64_MIN));
+    PyModule_AddObject(m, "INT64_MAX", PyLong_FromInt64(INT64_MAX));
+    PyModule_AddObject(m, "UINT64_MAX", PyLong_FromUInt64(UINT64_MAX));
```
This is not needed. You can define them in the Python code as `INT32_MAX = 2**31 - 1` etc.

vstinner replied:
I prefer to reuse <limits.h> C constants to avoid any typo.

serhiy-storchaka commented on Lib/test/test_capi/test_long.py:
```diff
@@ -185,25 +185,28 @@ def test_long_asint(self):
-        self.assertRaises(TypeError, aslong, b'2')
-        self.assertRaises(TypeError, aslong, '3')
-        self.assertRaises(SystemError, aslong, NULL)
+        self.check_long_asint(aslong, LONG_MIN, LONG_MAX)
```
This is nice. Could you use this also in other tests?

serhiy-storchaka replied:
I am creating a separate PR for this. It can be useful for other converters, and it is worth backporting.

serhiy-storchaka replied:
See #122113.

serhiy-storchaka reviewed (approved):
LGTM.

picnixz commented on Doc/c-api/long.rst:
```diff
@@ -337,6 +355,39 @@ distinguished from a number.  Use :c:func:`PyErr_Occurred` to disambiguate.
+.. c:function:: int PyLong_ToInt32(PyObject *obj, int32_t *value)
+                int PyLong_ToInt64(PyObject *obj, int64_t *value)
+
+   Return a signed C :c:expr:`int32_t` or :c:expr:`int64_t` representation of
```
I wouldn't say "Return" since you are not returning the representation. Maybe "Set *value* to the signed C representation of obj"? (I don't know if there are other functions like that already).

picnixz commented on Doc/c-api/long.rst:
```diff
@@ -337,6 +355,39 @@ distinguished from a number.  Use :c:func:`PyErr_Occurred` to disambiguate.
+.. c:function:: int PyLong_ToUInt32(PyObject *obj, uint32_t *value)
+                int PyLong_ToUInt64(PyObject *obj, uint64_t *value)
+
+   Return an unsigned C :c:expr:`uint32_t` or :c:expr:`uint64_t` representation
```
ditto

picnixz commented on Objects/longobject.c:
```diff
@@ -6681,3 +6681,67 @@ Py_ssize_t
+
+#define LONG_TO_INT(obj, value, type_name) \
+    do { \
+        int flags = Py_ASNATIVEBYTES_NATIVE_ENDIAN; \
```
```suggestion
		if (value == NULL) { \
			return -1; \
		} \
        int flags = Py_ASNATIVEBYTES_NATIVE_ENDIAN; \
```

I think that without this check, `PyLong_AsNativeBytes(obj, NULL, ...)` would segfault.

(Maybe this could be something to correct at the level of `PyLong_AsNativeBytes` though...)

vstinner replied:
I added "value must not be NULL" in the doc.

picnixz commented on Objects/longobject.c:
```diff
@@ -6681,3 +6681,67 @@ Py_ssize_t
+        int flags = (Py_ASNATIVEBYTES_NATIVE_ENDIAN \
+                     | Py_ASNATIVEBYTES_UNSIGNED_BUFFER \
+                     | Py_ASNATIVEBYTES_REJECT_NEGATIVE); \
+        Py_ssize_t bytes = PyLong_AsNativeBytes(obj, value, sizeof(*value), flags); \
```
Ditto.

vstinner committed:
Replace "To" with "As"

vstinner committed:
Merge branch 'main' into stdint

vstinner committed:
Use Py_ASNATIVEBYTES_ALLOW_INDEX

vstinner committed:
stable_abi: replace "To" with "As"

vstinner committed:
Address Bénédikt's review

serhiy-storchaka commented on Lib/test/test_capi/test_long.py:
```diff
@@ -631,6 +655,31 @@ def test_long_getsign(self):
+        as_uint32 = _testlimitedcapi.pylong_asuint32
+        from _testcapi import UINT32_MAX
+        self.check_long_asunsignedint(as_uint32, UINT32_MAX, use_index=True,
+                                      negative_value_error=True)
```
You can use `check_long_asint()`.
```suggestion
        self.check_long_asint(as_uint32, 0, UINT32_MAX,
                              negative_value_error=ValueError)
```

serhiy-storchaka reviewed (approved):
LGTM, but `check_long_asunsignedint` is now superseded by `check_long_asint`.

vstinner committed:
Remove check_long_asunsignedint() in tests

Reuse check_long_asint().

encukou commented on Doc/c-api/long.rst:
```diff
@@ -69,12 +69,30 @@ distinguished from a number.  Use :c:func:`PyErr_Occurred` to disambiguate.
+                PyObject* PyLong_FromInt64(int64_t value)
+
+   Return a new :c:type:`PyLongObject` object from a signed C
+   :c:expr:`int32_t` or :c:expr:`int64_t`, or ``NULL`` on failure.
```
```suggestion
   :c:expr:`int32_t` or :c:expr:`int64_t`, or ``NULL``
   with an exception set on failure.
```

encukou commented on Doc/c-api/long.rst:
```diff
@@ -69,12 +69,30 @@ distinguished from a number.  Use :c:func:`PyErr_Occurred` to disambiguate.
+                PyObject* PyLong_FromUInt64(uint64_t value)
+
+   Return a new :c:type:`PyLongObject` object from an unsigned C
+   :c:expr:`uint32_t` or :c:expr:`uint64_t`, or ``NULL`` on failure.
```
```suggestion
   :c:expr:`uint32_t` or :c:expr:`uint64_t`, or ``NULL``
   with an exception set on failure.
```

vstinner committed:
Update Doc/c-api/long.rst

Co-authored-by: Petr Viktorin <encukou@gmail.com>

vstinner committed:
Update Doc/c-api/long.rst

Co-authored-by: Petr Viktorin <encukou@gmail.com>
